### PR TITLE
Take and populate screenshot

### PR DESCRIPTION
**Pull request by adamfweidman**
Changes:
- getting and passing request headers check-id and execution-id to the runBrokenLinks() function -- handlers.ts and handlers.spec.ts
- shouldTakeScreenshot() function added.
- Some updates to uploadScreenshotToGCS(), mainly refactored out logic into: getStoragePathToExecution()
- add logic to increment number in screenshot_<number> when file to GCS.

### CODE REVIEW

File: packages/synthetics-sdk-broken-links/src/broken_links.ts [hunk 1]
```diff
@@ -121,18 +131,21 @@ export async function runBrokenLinks(
...
+      storageClient,
+      options.screenshot_options!.storage_location,
+      errors
+    );
 
-    // Create Promise and variables used to set and resolve the time limit
-    // imposed by `adjusted_synthetic_timeout`
-    const [timeLimitPromise, timeLimitTimeout, timeLimitresolver] =
-      getTimeLimitPromise(startTime, adjusted_synthetic_timeout_millis);
+    // TODO
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/src/broken_links.ts:141]: nit todo

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 2]
```diff
@@ -303,6 +335,37 @@ export function sanitizeObjectName(
...
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+) {
+  try {
+    const storageLocation = options.screenshot_options!.storage_location;
+    let writeDestination = '';
+
+    // extract folder name for a given storage location. If there is no '/'
+    // present then the storageLocation is just a folder
+    const firstSlashIndex = storageLocation.indexOf('/');
+    if (firstSlashIndex !== -1)
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:349]: nit consistency with the curly braces

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 3]
```diff
@@ -303,6 +335,37 @@ export function sanitizeObjectName(
...
+      writeDestination = storageLocation.substring(firstSlashIndex + 1);
+
+    // Ensure writeDestination ends with a slash for proper path joining
+    if (writeDestination && !writeDestination.endsWith('/')) {
+      writeDestination += '/';
+    }
+
+    writeDestination = path.join(
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:357]: Do we want a slash between checkId and executionId?

File: packages/synthetics-sdk-broken-links/test/example_html_files/integration_server.js [hunk 4]
```diff
@@ -14,6 +14,9 @@
 
 // Standard Libraries
 const path = require('path');
+// const proxyquire = require('proxyquire');
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/example_html_files/integration_server.js:17]: nit cleanup?

File: packages/synthetics-sdk-broken-links/test/unit/broken_links.spec.ts [hunk 5]
```diff
@@ -299,97 +381,10 @@ describe('runBrokenLinks', async () => {
...
-      expect(link.target_uri.endsWith(expectedTargeturis[index]));
-    });
+    broken_links_result?.followed_link_results?.forEach(
+      (link: BrokenLinksResultV1_SyntheticLinkResult, index: number) => {
+        expect(link.target_uri.endsWith(expectedTargeturis[index]));
+      }
+    );
   }).timeout(150000);
-
-  it('Completes a full passing execution', async () => {
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/unit/broken_links.spec.ts:307]: Maybe I missed this getting moved somewhere else, is there a reason this test got deleted?

File: packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts [hunk 6]
```diff
@@ -214,31 +235,125 @@ describe('GCM Synthetics Broken Links Utilies', async () => {
...
+
+      it('should return true when passed is false', () => {
+        const options = {
+          screenshot_options: { capture_condition: ApiCaptureCondition.ALL },
+        } as BrokenLinksResultV1_BrokenLinkCheckerOptions;
+        const result = shouldTakeScreenshot(options, false);
+        expect(result).to.be.true;
+      });
+    });
+    describe('screenshot_condition: FAILING', () => {
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts:284]: nit line break

File: packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts [hunk 7]
```diff
@@ -214,31 +235,125 @@ describe('GCM Synthetics Broken Links Utilies', async () => {
...
+          screenshot_options: {
+            capture_condition: ApiCaptureCondition.FAILING,
+          },
+        } as BrokenLinksResultV1_BrokenLinkCheckerOptions;
+        const result = shouldTakeScreenshot(options, true);
+        expect(result).to.be.false;
+      });
+    });
+
+    describe('screenshot_condition: NONE', () => {
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts:306]: Nit on consistency here, if we can combine the NONE case into 1 test then we should do it for the others, or split up the NONE case into 2 tests

File: packages/synthetics-sdk-broken-links/test/unit/navigation_func.spec.ts [hunk 8]
```diff
@@ -74,6 +104,15 @@ describe('GCM Synthetics Broken Links Navigation Functionality', async () => {
     // Create a new page for each test
     page = await browser.newPage();
     page.setCacheEnabled(false);
+
+    // Stub the page.screenshot function
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/unit/navigation_func.spec.ts:108]: Overly critical nit but consistency with commenting and honestly some of these comments aren't needed, the Sinon.stub tells you what it's doing, Sinon.restore tells you it's restoring, these comments aren't in other places these are called, etc.

File: packages/synthetics-sdk-broken-links/test/unit/storage_func.spec.ts [hunk 9]
```diff
@@ -163,96 +161,118 @@ describe('GCM Synthetics Broken Links storage_func suite testing', () => {
...
+        );
+
+        expect(result2.screenshot_file).to.equal('screenshot_2.png');
+        expect(result2.screenshot_error).to.deep.equal({});
 
-      expect(result.screenshot_file).to.equal('');
-      expect(result.screenshot_error).to.deep.equal({
-        error_type: 'StorageFileUploadError',
-        error_message: `Failed to upload screenshot for ${filename}. Please reference server logs for further information.`,
+        // TODO
```
Comment (jdurelli) [inline packages/synthetics-sdk-broken-links/test/unit/storage_func.spec.ts:215]: nit remove

**Review by jdurelli — APPROVED** [review]

File: packages/synthetics-sdk-broken-links/src/link_utils.ts (hunk 3 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:357]: yup!

Comment (jdurelli) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:358]: I guess this question was getting at - if the path.join function adds a slash by are we doing it manually for write destination above?

Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:358]: oh I see, yes I think that may be redundant

**Pull request merged**